src/components/button/Button.js

```diff
@@ -1,30 +1,160 @@
+                    <span>{ children }</span>
+                </button>
+                <ProgressBar
+                    running={ loading }
+                    className={ styles.progressBar }
+                    onBegin={ this.handleProgressBarBegin }
+                    onEnd={ this.handleProgressBarEnd } />
+
+                <span className={ successBlockClassName } onTransitionEnd={ this.handleSpanTransitionEnd }>
+                    <CheckmarkIcon className={ styles.checkmark } onTransitionEnd={ this.handleSuccessIconTransitionEnd } />
+                </span>
+                <span className={ errorBlockClassName } onTransitionEnd={ this.handleSpanTransitionEnd }>
+                    <CrossIcon className={ styles.cross } onTransitionEnd={ this.handleErrorIconTransitionEnd } />
+                </span>
+            </div>
         );
     }
+
+    maybeHandleFeedbackChange(prevFeedback) {
+        const { feedback } = this.props;
+
+        // Skip if `feedback` hasn't changed
+        if (feedback === prevFeedback) {
+            return;
+        }
+
+        // If feedback prop changed to `success` or `error` without passing through `loading`,
+        // force the intermidate `loading` state
+        if ((feedback === 'success' || feedback === 'error') && this.state.feedback !== 'loading') {
+            this.setState({ feedback: 'loading' }, () => {
+                this.setState({ feedback });
+            });
+        // Otherwise, simply copy the feedback to the state
+        } else {
+            this.setState({ feedback });
+        }
+    }
+
+    clearFeedbackOutcomeTimers() {
+        clearTimeout(this.resetFeedbackOutcomeTimeoutId);
+    }
+
+    handleProgressBarBegin = () => {
+        this.clearFeedbackOutcomeTimers();
+        this.setState({ feedbackOutcome: null });
+    };
+
+    handleProgressBarEnd = () => {
+        const { feedback } = this.props;
+
+        this.clearFeedbackOutcomeTimers();
+        this.setState({ feedbackOutcome: feedback }, () => {
+            this.clearFeedbackOutcomeTimers();
+            this.resetFeedbackOutcomeTimeoutId = setTimeout(() => {
+                this.setState({ feedback: 'none', feedbackOutcome: null });
+            }, FEEDBACK_OUTCOME_VISIBLE_DURATION);
+        });
+    };
+
+    handleSpanTransitionEnd = (event) => {
+        event.stopPropagation();
+        if (this.enterAnimation) {
+            this.enterAnimation = false;
+
+            return;
+        }
+        this.enterAnimation = true;
+    };
+
+    handleIconTransitionEnd = (event, isSuccess) => {
+        event.stopPropagation();
+        if (this.enterAnimation) {
+            this.props.onAnimationEnd && this.props.onAnimationEnd(isSuccess);
+        }
+    };
+
+    handleSuccessIconTransitionEnd = (e) => this.handleIconTransitionEnd(e, true);
+
+    handleErrorIconTransitionEnd = (e) => this.handleIconTransitionEnd(e, true);
```

Comment:
As far as I understood what you're doing here, in this case, `this.handleIconTransitionEnd` must be called with its second argument as `false` instead of `true`. Am I right?

Reply:
Yes, nice catch 👍